File: src/main/java/bad/robot/radiate/monitor/ScheduledMonitor.java
package bad.robot.radiate.monitor;

import java.util.List;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;

import static java.util.concurrent.TimeUnit.SECONDS;
import static java.util.stream.Collectors.toList;
import static java.util.stream.StreamSupport.stream;

public class ScheduledMonitor implements Monitor {

    private final ScheduledExecutorService executor;
    private final int frequency = 30;

    public ScheduledMonitor(ScheduledExecutorService executor) {
        this.executor = executor;
    }

    @Override
    public Iterable<ScheduledFuture<?>> start(Iterable<MonitoringTask> tasks) {
        return stream(tasks.spliterator(), false).map(task -> executor.scheduleWithFixedDelay(task, 0, frequency, SECONDS)).collect(toList());
    }

    @Override
    public void cancel(Iterable<ScheduledFuture<?>> tasks) {
        tasks.forEach(task -> task.cancel(true));
    }

    @Override
    public List<Runnable> stop() {
        List<Runnable> tasks = executor.shutdownNow();
        tasks.stream().forEach(task -> ((ScheduledFuture<?>) task).cancel(true));
        return tasks;
    }
}

File: src/main/scala/bad/robot/radiate/monitor/ScheduledMonitor.scala
package bad.robot.radiate.monitor

import java.util.concurrent.TimeUnit.SECONDS
import java.util.concurrent.{ScheduledExecutorService, ScheduledFuture}

import scala.collection.JavaConverters._

class ScheduledMonitorS(executor: ScheduledExecutorService) extends MonitorS {

  private val frequency: Int = 30

  def start(tasks: List[MonitoringTask]): List[ScheduledFuture[_]] = {
    tasks.map(task => executor.scheduleWithFixedDelay(task, 0, frequency, SECONDS))
  }

  def cancel(tasks: List[ScheduledFuture[_]]) {
    tasks.foreach(task => task.cancel(true))
  }

  def stop: List[Runnable] = {
    val tasks = executor.shutdownNow.asScala
    tasks.foreach(task => task.asInstanceOf[ScheduledFuture[_]].cancel(true))
    tasks.toList
  }

}